tests/test_flatAndIlluminationCorrection.py

```diff
@@ -98,6 +98,30 @@ def testIllum2(self):
     def testIllum3(self):
         self.doIllum(scaling=3.7)
 
+    def testGainAndReadnoise(self):
+        import lsst.afw.image as afwImage
+        from lsst.afw.cameraGeom.testUtils import DetectorWrapper
+        from lsst.ip.isr import IsrTask
+
+        isrTask = IsrTask()
+
+        detector = DetectorWrapper().detector
+        raw = afwImage.ExposureF(detector.getBBox())
+
+        level = 10
+        readNoise = 1
+        raw.image.set(level)
+
+        amp = detector[0]
+        amp.setReadNoise(readNoise)
+
+        for gain in [-1, 0, 0.1, 1]:
+            amp.setGain(gain)
+            isrTask.updateVariance(raw, amp)
+            if gain <= 0:
+                gain = 1
+
+        self.assertEqual(raw.variance.get(0, 0), level/gain + readNoise**2)
```

Comment:
I see three possible branches in the code you added: `gain is nan`, `gain <= 0`, and `gain > 0`. There is only one assert here, so it can only test one of those branches.

Separately, why are you doing `raw.variance.get(0, 0)`, instead of checking all values of `raw.variance`?

Reply:
The line's indentation is incorrect; it should be in the loop.  I'm not sure how this happened as I checked that the code failed when I reverted the fix.  User error.

I only need to check one pixel as they're all the same, so this is simpler and equivalent.

Reply:
1. That still doesn't test the nan case.

2. How do you know the code is treating all the pixels correctly unless you actually test all the pixels?

Reply:
What NaN case? 

I'm adding a test to check that we now handle gains <= 0, so I don't need to test all the pixels -- it wouldn't gain anything.

And I don't understand your question about the readNoise.  I'm adding a test that we handle an illegal value from the cameraGeom data, not a test to check that we handle variances correctly.  I would like better test coverage for the IsrTask, but that should be calling the IsrTask, not internals.

Reply:
The code itself has an `if isnan(gain)`, which means that we need a test case for gain being `NaN`.

Again, how do you know that all of the pixels are treated correctly unless you test them?

The method you are testing is called `updateVariance`. There is a term related to `readNoise**2`. `1` doesn't give you much information when you square it.

Reply:
A value 1.5 is now used for read noise.  A test for a NaN value of gain has been added.  Concerning the question about testing all pixels, I agree with Robert that this is outside the scope of this unit test.  This is a unit test for the handling of illegal values from the cameraGeom data. not the correct handling of variances.